### ISSUE TITLE
Would you like a default setup for tests?

### ISSUE THREAD

**Issue by chantastic** [JedWatson/generator-react-component#4]
@JedWatson 

Would you accept a PR that added a default setup for testing the component?

I'd like to see this library ship with a bare-minimum testing setup. I'm no testing expert but I've spent a significant amount of time working out a testing setup for our teams. I've found [Mocha](http://mochajs.org) and React [shallowRenderer](https://facebook.github.io/react/docs/test-utils.html#shallow-rendering) to be the best setup available today.
#### What it would add
- devDependencies
  - [babel](https://www.npmjs.com/package/babel)
  - [mocha](https://www.npmjs.com/package/mocha)
- package.json
  - `test` command
- mocha config
  - `mocha.opts` file with default command options
  - `test/utils/dom.js` with `document` mock and commented code for an browser setup with [jsdom](https://www.npmjs.com/package/jsdom)
#### Why Mocha over Jest?

There are 3 reasons I think Mocha is a great choice for testing in this project. None of these reasons will be a surprise to you but I'll list them for thoroughness:
1. **Extensibility** Mocha is an un-opinionated library. Assertion-style (assert/expect/should) can (and must be chosen by the developer. The same goes for stub/mock/spy libraries. Jest is much more prescriptive, being built atop Jasmine.
2. **Speed** Mocha is insanely fast. This generator is designed mostly for small projects, I'm assuming 1-10 components. Mocha should perform exceptional for this use case. Jest has a theoretical speed benefit for large projects, running tests in parallel. But I don't sees this every being an issue.
3. **Ergonomics** The command-line API for `Jest` is totally opaque. I am not aware of a way to easily `watch` the filesystem or run a single test via the command-line. This slows the feedback loop dramatically when BDDing a component. `Mocha`s command-line interface is accessible and well documented.
#### Shortcomings

One benefit that Jest has over Mocha is that it's being built alongside React. This becomes obvious in issues [like this](https://github.com/facebook/react/issues/4019) where tests fail due to an expectation in React for a `global.document` definition, despite the fact that `shallowRenderer` is designed for testing without a `document`.

Mocha makes it pretty simple circumvent these. But a small amount of  configuration is necessary.
#### What would you like to see?

I'd love to get started of this. Let me know if you think it would be a good fit for the project and what your criteria for acceptance would be.

:heart:

@chantastic

**Comment by JedWatson** [comment]
@chantastic I am _so keen_ for this, thank you. This project should _absolutely_ ship with testing config & examples built in.

Given the complexity of working in a generator project (i.e being able to develop, run and review changes to the templates) it might be easier to generate a simple project (e.g `github.com/chantastic/react-component-test-prototype`) that you can build the tests in, and we can review there and pull your changes across.

I'd like to pull @bruderstein in for commentary too, he's just put in a huge effort getting [react-select](https://github.com/JedWatson/react-select/graphs/contributors) under test. There's quite a lot that's gone into setting up those tests that might be worth bringing across.

We also moved from `Jest` to `mocha` not too long ago, it was an interesting process with competing PRs contributed for both (see JedWatson/react-select#247). So I think mocha is the right call to add here.

One problem I've had in `react-select` is we need to keep `jsdom` at `^3.1.2` because later releases require `iojs` (and won't work with `node`) which, while I'm sure they have good reasons for it, I find too opinionated for an open source component with lots of contributors.

I'm not sure that `shallowRenderer` would work for comprehensive tests like `react-select` has due to its limitations, but I haven't looked into it too closely.

For future versions of this generator, I've been thinking about getting an equivalent web pack config set up, so we can make it an option to use web pack or browserify for development (even now I still have a preference for browserify, but I know the value of webpack + hot loader, especially for component development). Maybe we could (if it's not _too_ much code to maintain) end up doing the same for mocha / jest as a test-suite selection.

The keystone generator offers several similar choices (e.g. we have the same templates available in four different languages, and you can choose less or sass) and it's been a really popular thing to provide.

### TL;DR

No strict criteria for acceptance, I trust your judgement and code and am very happy for you to be taking this on :grinning:

Looking forward to seeing what you come up with!

**Comment by chantastic** [comment]
Excellent!

> ...it might be easier to generate a simple project

I've setup a branch, [mocha-tests](https://github.com/chantastic/react-media-object/tree/mocha-tests) on my React-Media-Object. I can setup a project with a more robust domain. In the meantime, this is a real-world component I had on hand.

You can see the [full comparison](https://github.com/chantastic/react-media-object/pull/4). There is very little required to get a minimal testing setup. I made the history as "step-able" as possible.

> We also moved from Jest to mocha not too long ago...

My experience with `Jest` was very bad. It ships with a copy-pasted version of `Jasmine 1.3` and depends on `jsdom` which is an EXTREMELY opinionated library with (as you mentioned) no support for `node`. I found it to be far more trouble than it was worth (at this point).

> I'm not sure that shallowRenderer would work for comprehensive tests like react-select has due to its limitations, but I haven't looked into it too closely.

It's very possible that it wouldn't. That said, I think shipping with `shallowRenderer` might be a really great "un-opinionated default". Even in it's current iteration, it covers a good number of testing scenarios. Using that `document` trick, I've able to test anything events and state changes without issue.

React-Media-Object doesn't have any events but I included an example of how they can be tested with `shallowRenderer`.

> I've been thinking about getting an equivalent web pack config set up, so we can make it an option to use web pack or browserify... Maybe we could end up doing the same for mocha / jest as a test-suite selection.

That sounds great!

I'm curious to see what peoples preference is on this. Given how small these components are likely to be, I'm not sure how end-devs will fully realize the benefits of auto-mocking and parallel test-running.

---

Let me know your thoughts. Right now, I'm trying to keep it as minimal and extensible as possible. I'm happy to pin it down a little more if you think there's room for that—possibly rallying behind `chai` (assertions) and `sinon` (spies/stubs) as default includes.

**Comment by chantastic** [comment]
Hi @bruderstein, I think we met at React Europe :)

**Comment by JedWatson** [comment]
> I think shipping with shallowRenderer might be a really great "un-opinionated default"

:+1:

**Comment by bruderstein** [comment]
Hi,

Not sure I can add much to this, I think I agree with pretty much all of the points made. 

My experience of jest was also bad, although admittedly a while ago and I haven't taken a fresh look since. I do have a more fundamental dislike of it though - I'd rather not have a metric tonne of magic happening when I'm trying to prove that something works.

> I'm not sure that shallowRenderer would work for comprehensive tests like react-select has due to its limitations, but I haven't looked into it too closely.

_(Disclaimer: I've never successfully used shallowRenderer)_
I think for a lot of cases it would, but could limit the refactoring possibilities (without changing tests) - currently, for example, as long as the label is rendered in an element with the right class (_insert your own witty comment about inline styles here_), the test is happy, which is actually what the user wants too. If we were using `shallowRenderer`, I think we'd have to care _how_ it got rendered (e.g. `<Value ... />` or directly, etc.).

As an end user, I'd like to see how to write tests using a `jsdom`, with `TestUtils.Simulate`,  `findDOMNode()` and friends - could that be an option?

I find being able to test the DOM is conceptually simpler, even though it often isn't simpler in practice. Testing `onClick()` directly doesn't test the event is bound correctly - I know this is unlikely, but it's the sort of typo-type problem that's great when a test catches. Or is there a way to do that?

> I'm happy to pin it down a little more if you think there's room for that—possibly rallying behind chai (assertions) and sinon (spies/stubs) as default includes.

I'm gonna sing about [unexpected](https://github.com/unexpectedjs/unexpected) and [unexpected-dom](https://github.com/Munter/unexpected-dom) here.  This would be another nice option to have. :+1: for sinon as a default.

> Hi @bruderstein, I think we met at React Europe :)

Hi, we did!

**Comment by chantastic** [comment]
@bruderstein that is such great feedback! thank you :)

I created a new PR (into my sample project) that shows just the difference between what's needed for a basic `shallowRenderer` setup and a `jsdom` setup.

Here it is: https://github.com/chantastic/react-media-object/pull/5/files

**In both setups the the `shallowRenderer` version works**. It seems to me that the most pragmatic (and developer-friendly) solution is ship with `jsdom`. This way the developer can hit the ground running with either preference.

The only hold-up is that it feels kinda dirty shipping with an unsupported library version. I think, in this case, it might be understandable. But who knows :)

**Comment by JedWatson** [comment]
> The only hold-up is that it feels kinda dirty shipping with an unsupported library version

The biggest way that this has hit me is I often go through updating the dependencies of my projects. A few times I've accidentally updated jsdom and it breaks all my tests. This could be a real catch for developers who've just used the generator and aren't familiar with the situation, especially because you can't include comments in package.json explaining it to them :disappointed:

**Comment by chantastic** [comment]
> This could be a real catch for developers who've just used the generator and aren't familiar with the situation

Agreed. This would be a very frustrating introduction.

**Should we punt on `jsdom` until after the `io.js/node` merge?**

> I find being able to test the DOM is conceptually simpler, even though it often isn't simpler in practice.

I think this is very true. `shallowRenderer` is difficult because it doesn't allow you to borrow from any of your existing testing experience.

#### so, where are we?

I've added another PR that adds all of the `jsdom` code as a comment, with a mention in (what would be) the generated test file. Let me know if you think that would be a good middle ground until `jsdom` settles out.

https://github.com/chantastic/react-media-object/pull/6/files

**Comment by bruderstein** [comment]
> I've added another PR that adds all of the jsdom code as a comment, with a mention in (what would be) the generated test file. Let me know if you think that would be a good middle ground until jsdom settles out.
> 
> https://github.com/chantastic/react-media-object/pull/6/files

I think this is a really, really good middleground. It's very clear what's going on, how to add DOM rendering, and uses shallowRenderer as a default. Brilliant!

The only thing that may be worth doing, is to add a check in `test/utils/document.js` if global.window is already defined, to not overwrite it. This doesn't affect most situations, but running tests with re-used runner processors can mean that the `document` React uses, is not the same as the current `document`, which means you get a jsdom error about the `document` not being the right one.  

This happens with [wallaby.js](http://wallabyjs.com) for example - it took me a while to figure out why, and what the best solution was - simply not recreating it if it's already there works reliably and is easy to understand. See the much-less-professionally-written [jsdomHelper.js](https://github.com/JedWatson/react-select/blob/master/testHelpers/jsdomHelper.js) in react-select.  (So gonna be stealing your test/utils/document.js file ... :wink:)

**Comment by chantastic** [comment]
> if global.window is already defined, to not overwrite it

Super smart. Thanks for knowing that :)

> (So gonna be stealing your test/utils/document.js file ... :wink:)

Please do! I stole it from from Jake Trent: http://jaketrent.com/post/testing-react-with-jsdom/

**Comment by chantastic** [comment]
@JedWatson 

> A few times I've accidentally updated jsdom and it breaks all my tests

I'm showing my lack of familiarity with node. But is it possible (as with Ruby/Bundler), to set an upper limit?

`"jsdom": ">3.0 <4.0"`

**Comment by JedWatson** [comment]
@chantastic npm uses semver so it's not a problem when you install, just when you're checking for newer versions of dependencies you might need to update to.

I use this great package called [npm-check-updates](https://www.npmjs.com/package/npm-check-updates) that tells you which packages have updates available, which version is available, and can automate the changes to `package.json`.

For me, having out-of-date dependencies is a bit like having linter errors. They aren't necessarily a problem, but everything isn't "green" :)

Whenever I've got jsdom in a project and I use npm-check-updates to upgrade its dependencies, I need to go back and manually downgrade jsdom again.

Not the end of the world, and not necessarily worse than not using jsdom, but now you know the background!

**Comment by chantastic** [comment]
@JedWatson that makes sense. in the `bundler` world, you're never allowed to upgrade past your max version, without explicitly changing your `gemfile`.

If that's the case, I might make a run at the **mocha + comments** approach, possibly with the addition of a wiki entry.

**Comment by jeffbski** [comment]
@JedWatson I am not familiar with npm-check-updates, though I usually just use the built in 
`npm outdated --depth=0` which shows me current version, highest in my semver, and latest version numbers.

Then I just do 
`npm update FOO --save` (or --save-dev for dev deps) to update within the package.json semver and write updated version to package.json 

OR 

if I want to jump to latest 
`npm install FOO@latest --save` to jump to the latest version (regardless of package.json version) and write to package.json.

I am guessing from the readme that npm-check-updates makes this easier to do things in bulk? Any other differences from the pure npm approach?

Thanks.

**Comment by bruderstein** [comment]
@chantastic - just playing with the shallowRenderer, and I'm not sure I follow the event example, at least, I can't see how the onClick ends up in the props, and if it was there, what we could do with it, as the renderer gives back a pretty raw object, so testing state changes doesn't seem doable.

What I am I missing?

**Comment by chantastic** [comment]
@bruderstein it's totally possible that my example doesn't work, give that nothing in that component changes state. But it's something I do with our application components all the time. 

Here's the example in the ReactJS source that I borrowed from:

https://github.com/facebook/react/blob/7deab28347eaf206217ffee8e4475a1935b02cb8/src/test/__tests__/ReactTestUtils-test.js#L88-L140

I can build up an isolated test case if that would be helpful. Sorry I fell off this issue. Hoping to put some time at it tomorrow :)

**Comment by bruderstein** [comment]
@chantastic awesome, thank you, now I understand. I don't know why I didn't think to look at the react tests for the shallow renderer!
